### PR TITLE
Sp 07 wed

### PR DESCRIPTION
**Pull request by fyarkiy**

### CODE REVIEW

File: src/main/java/com/cinema/config/validations/EmailValidator.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+            + "[^-][A-Za-z0-9-]+(\\.[A-Za-z0-9-]+)*(\\.[A-Za-z]{2,})$";
+
+    @Override
+    public void initialize(EmailConstraint constraint) {
+    }
+
+    @Override
+    public boolean isValid(String emailField, ConstraintValidatorContext context) {
+
+        return emailField.matches(REGEX);
```
Comment (jah701) [inline src/main/java/com/cinema/config/validations/EmailValidator.java]: Maybe we can add null check here :)

File: src/main/java/com/cinema/config/validations/EmailValidator.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+public class EmailValidator implements ConstraintValidator<EmailConstraint, String> {
+    private static final String REGEX = "^(?=.{1,64}@)[A-Za-z0-9_-]+(\\.[A-Za-z0-9_-]+)*@"
+            + "[^-][A-Za-z0-9-]+(\\.[A-Za-z0-9-]+)*(\\.[A-Za-z]{2,})$";
+
+    @Override
+    public void initialize(EmailConstraint constraint) {
```
Comment (jah701) [inline src/main/java/com/cinema/config/validations/EmailValidator.java]: We can skip this method

File: src/main/java/com/cinema/config/validations/PasswordMatchConstraint.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    String message() default "Passwords are not the same";
+
+    String password();
+
+    String repeatPassword();
+
+    Class<?>[] groups() default {};
+    Class<? extends Payload>[] payload() default {};
+
+    /*@Target({ElementType.TYPE})
```
Comment (jah701) [inline src/main/java/com/cinema/config/validations/PasswordMatchConstraint.java]: Mentors will make you remove it :)

File: src/main/java/com/cinema/config/validations/PasswordMatchValidator.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    private String repeatPassword;
+
+    public void initialize(PasswordMatchConstraint constraint) {
+        this.password = constraint.password();
+        this.repeatPassword = constraint.repeatPassword();
+    }
+
+    public boolean isValid(UserRequestDto userRequestDto, ConstraintValidatorContext context) {
+
+        Object passwordValue = new BeanWrapperImpl(userRequestDto)
```
Comment (jah701) [inline src/main/java/com/cinema/config/validations/PasswordMatchValidator.java]: According our qna with Dima should we cast it to String? I have cast it to String, not sure if it is correct. Just double check :)

File: src/main/java/com/cinema/controllers/ShoppingCartController.java [hunk 5]
```diff
@@ -31,14 +33,16 @@ public ShoppingCartController(ShoppingCartService shoppingCartService,
 
     @PostMapping("/movie-sessions")
     public void addMovies(@RequestParam Long movieSessionId,
-                          @RequestParam Long userId) {
+                          Authentication authentication) {
+        UserDetails principle = (UserDetails) authentication.getPrincipal();
         shoppingCartService.addSession(movieSessionService.getById(movieSessionId),
-                userService.getById(userId));
```
Comment (jah701) [inline src/main/java/com/cinema/controllers/ShoppingCartController.java:36]: I have also forgotten to delete getById() method as we dont use it :)

**Review by jah701 — APPROVED** [review]
Very well done! :)

File: src/main/java/com/cinema/config/validations/EmailConstraint.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
+package com.cinema.config.validations;
```
Comment (KseniiaMakarova) [inline src/main/java/com/cinema/config/validations/EmailConstraint.java]:
```suggestion
package com.cinema.validation;
```
same for both your annotations and validators

File: src/main/java/com/cinema/config/validations/EmailValidator.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+import javax.validation.ConstraintValidatorContext;
+
+public class EmailValidator implements ConstraintValidator<EmailConstraint, String> {
+    private static final String REGEX = "^(?=.{1,64}@)[A-Za-z0-9_-]+(\\.[A-Za-z0-9_-]+)*@"
+            + "[^-][A-Za-z0-9-]+(\\.[A-Za-z0-9-]+)*(\\.[A-Za-z]{2,})$";
+
+    @Override
+    public boolean isValid(String emailField, ConstraintValidatorContext context) {
+
+        return emailField.matches(REGEX);
```
Comment (KseniiaMakarova) [inline src/main/java/com/cinema/config/validations/EmailValidator.java]: I agree, let's add null check as well

File: src/main/java/com/cinema/config/validations/PasswordMatchValidator.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    public boolean isValid(UserRequestDto userRequestDto, ConstraintValidatorContext context) {
+
+        String passwordValue = (String) new BeanWrapperImpl(userRequestDto)
+                .getPropertyValue(password);
+        String repeatPasswordValue = (String) new BeanWrapperImpl(userRequestDto)
+                .getPropertyValue(repeatPassword);
+
+        return passwordValue == null || passwordValue.equals(repeatPasswordValue);
```
Comment (KseniiaMakarova) [inline src/main/java/com/cinema/config/validations/PasswordMatchValidator.java]: what if password is null, and repeatPassword is not null?

File: src/main/java/com/cinema/model/dto/OrderResponseDto.java [hunk 9]
```diff
@@ -1,8 +1,10 @@
 package com.cinema.model.dto;
 
 import java.util.List;
+import javax.validation.constraints.NotNull;
 
 public class OrderResponseDto {
+    @NotNull
```
Comment (KseniiaMakarova) [inline src/main/java/com/cinema/model/dto/OrderResponseDto.java]: this does nothing for response dtos, let's remove it

File: src/main/java/com/cinema/service/UserService.java [hunk 10]
```diff
@@ -8,5 +8,5 @@
 
     Optional<User> findByEmail(String email);
 
-    User getById(Long id);
+    /*User getById(Long id);*/
```
Comment (KseniiaMakarova) [inline src/main/java/com/cinema/service/UserService.java]: fix this

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]
also, please remove salt from User model

File: src/main/java/com/cinema/config/validations/PasswordMatchValidator.java (hunk 8 above)
Comment (fyarkiy) [reply src/main/java/com/cinema/config/validations/PasswordMatchValidator.java]: I have NotNull on this field in the Dto - so I have a response "password can't be null". A message that those two "are not equal" will not appear but in any case we have an error which should be fixed.

File: src/main/java/com/cinema/model/dto/ShoppingCartResponseDto.java [hunk 11]
```diff
@@ -1,9 +1,12 @@
 package com.cinema.model.dto;
 
 import java.util.List;
+import javax.validation.constraints.NotNull;
 
 public class ShoppingCartResponseDto {
+    @NotNull
```
Comment (KseniiaMakarova) [inline src/main/java/com/cinema/model/dto/ShoppingCartResponseDto.java]: remove this as well

File: src/main/java/com/cinema/model/dto/TicketResponseDto.java [hunk 12]
```diff
@@ -1,7 +1,11 @@
 package com.cinema.model.dto;
 
+import javax.validation.constraints.NotNull;
+
 public class TicketResponseDto {
+    @NotNull
```
Comment (KseniiaMakarova) [inline src/main/java/com/cinema/model/dto/TicketResponseDto.java]: same

File: src/main/java/com/cinema/validations/PasswordMatchValidator.java [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    public boolean isValid(UserRequestDto userRequestDto, ConstraintValidatorContext context) {
+
+        String passwordValue = (String) new BeanWrapperImpl(userRequestDto)
+                .getPropertyValue(password);
+        String repeatPasswordValue = (String) new BeanWrapperImpl(userRequestDto)
+                .getPropertyValue(repeatPassword);
+
+        return passwordValue == null || passwordValue.equals(repeatPasswordValue);
```
Comment (KseniiaMakarova) [inline src/main/java/com/cinema/validations/PasswordMatchValidator.java]:
```suggestion
        return passwordValue != null && passwordValue.equals(repeatPasswordValue);
```
we can not be sure that @NotNull will work before our custom validation

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**